src/Castle.Core/DynamicProxy/ProxyGenerationOptions.cs

```diff
@@ -204,7 +208,53 @@ public override int GetHashCode()
 			result = 29*result + (Selector != null ? 1 : 0);
 			result = 29*result + MixinData.GetHashCode();
 			result = 29*result + (BaseTypeForInterfaceProxy != null ? BaseTypeForInterfaceProxy.GetHashCode() : 0);
+		    result = 29*result + GetContentsHashCode(AdditionalAttributes);
 			return result;
 		}
+
+        private static int GetContentsHashCode<T>(IList<T> list)
+        {
+            var result = 0;
+            for (var i = 0; i < list.Count; i++)
+            {
+                if (list[i] != null)
+                {
+                    // simply XOR since order does not matter
+                    result ^= list[i].GetHashCode();
+                }
+            }
+
+            return result;
+        }
+
+        private static bool ListsEquivalent<T>(IList<T> listA, IList<T> listB)
+        {
+            if (listA.Count != listB.Count)
+            {
+                return false;
+            }
+
+            // order is not important, just make sure that each entry in A is also found in B

```

Comment:
You are right, in fact if the same attribute is there an even number of times, it will effectively be XOR'd out which would result in the HashCode not taking it into account at all.

I'll adjust the algorithm to group equal HashCodes together and alter the "group hash" which will be based on the number of equal elements. This way, two types that have the same repeated attributes would only have the same hash if they had the same number (order would still not matter).

How would you feel about comparing HashCodes in Equals before doing comparisons for each object individually? If the HashCodes match, we can inspect each. If not, then we won't iterate through the array.